lib/APPTurtle/src/App.cpp

```diff
@@ -234,6 +220,126 @@ void App::fatalErrorHandler()
     }
 }
 
+bool App::setupMicroRosClient()
+{
+    bool isSuccessful = false;
+
+    /* Subscriber for Geometry Twist ROS messages. */
+    typedef Subscriber<geometry_msgs__msg__Twist> CmdVelSubscriber;
+
+    /* Create the Subscriber Callback. */
+    CmdVelSubscriber::RosTopicCallback twistCallback = [this](const geometry_msgs__msg__Twist* msgData)
+    {
+        if (nullptr == msgData)
+        {
+            LOG_ERROR("TwistCallback received nullptr.");
+        }
+        else
+        {
+            LOG_DEBUG("Twist data received.");
+
+            /* Short blink to indicate reception. */
+            Board::getInstance().getBlueLed().enable(true);
+
+            m_turtleSpeedSetpoint      = *msgData;
+            m_isNewTurtleSpeedSetpoint = true;
+
+            Board::getInstance().getBlueLed().enable(false);
+        }
+    };
+
+    /* Create instance of CmdVelSubscriber. Will be deleted by the MicroRosClient. */
+    CmdVelSubscriber* twistSub = new (std::nothrow)
+        CmdVelSubscriber(TOPIC_NAME_CMD_VEL, ROSIDL_GET_MSG_TYPE_SUPPORT(geometry_msgs, msg, Twist), twistCallback);
+
+    /* Register the subscriber. */
+    if (nullptr == twistSub)
+    {
+        LOG_ERROR("Could not create instance of CmdVelSubscriber");
+    }
+    else if (false == m_ros.registerSubscriber(twistSub))
+    {
+        LOG_ERROR("Could not register the CmdVelSubscriber.");
+    }
+    else
+    {
+        isSuccessful = true;
+    }
+
+    return isSuccessful;
+}
+
+bool App::setupSerialMuxProtServer()
+{
+    bool isSuccessful = false;
+
+    m_serialMuxProtChannelIdStatus = m_smpServer.createChannel(STATUS_CHANNEL_NAME, STATUS_CHANNEL_DLC);
+    m_serialMuxProtChannelIdMotorSpeeds =
+        m_smpServer.createChannel(SPEED_SETPOINT_CHANNEL_NAME, SPEED_SETPOINT_CHANNEL_DLC);
+
+    if ((0U == m_serialMuxProtChannelIdStatus) || (0U == m_serialMuxProtChannelIdMotorSpeeds))
+    {
+        LOG_ERROR("Failed to create SerialMuxProt channels.");
+    }
+    else
+    {
+        isSuccessful = true;
+        m_statusTimer.start(STATUS_TIMER_INTERVAL);
+    }
+
+    return isSuccessful;
+}
+
+void App::handleTurtle()
+{
+    /* Check for new data. */
+    if (true == m_isNewTurtleSpeedSetpoint)
+    {
+        m_isNewTurtleSpeedSetpoint = false;
+        LOG_DEBUG("New turtle speed setpoint received. %f", m_turtleSpeedSetpoint.linear.x);
+        SpeedData payload;
+
+        if (m_turtleSpeedSetpoint.linear.x > 0.0F)
+        {
+            payload.left  = 3000;
```

Comment:
csn you explain the number value and why you not consider what x or z is ?

Reply:
Will add some comments, just for this PR. A rework of the app and a state machine are probably necessary to retrieve the maximum motor speed and adjust the setpoint accordingly to take the input of x and z into account